phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexUpgradeTool.java

```diff
@@ -0,0 +1,529 @@
+            if (upgrade) {
+                if (!admin.getDescriptor(TableName.valueOf(indexName))
+                        .hasCoprocessor(GlobalIndexChecker.class.getName())) {
+                    if (!dryRun) {
+                        indexTableDescBuilder.addCoprocessor(GlobalIndexChecker.class.getName(),
+                                null, QueryServicesOptions.DEFAULT_COPROCESSOR_PRIORITY, prop);
+                    }
+                    LOGGER.info("Loaded GlobalIndexChecker coprocessor on index table " + indexName);
+                } else {
+                    LOGGER.info("GlobalIndexChecker coprocessor on index table " + indexName + "is already loaded");
+                }
+            } else {
+                if (admin.getDescriptor(TableName.valueOf(indexName))
+                        .hasCoprocessor(GlobalIndexChecker.class.getName())) {
+                    if (!dryRun) {
+                        indexTableDescBuilder.removeCoprocessor(GlobalIndexChecker.class.getName());
+                    }
+                    LOGGER.info("Unloaded GlobalIndexChecker coprocessor on index table "+indexName);
+                } else {
+                    LOGGER.info("GlobalIndexChecker coprocessor on index table " + indexName + "is already unloaded");
+                }
+            }
+            if (!dryRun) {
+                admin.modifyTable(indexTableDescBuilder.build());
+            }
+        }
+    }
+
+
+    private void rebuildIndexes(String dataTable, Configuration conf) {
+        String schema = SchemaUtil.getSchemaNameFromFullName(dataTable);
+        String table = SchemaUtil.getTableNameFromFullName(dataTable);
+        for(String index: rebuildMap.get(dataTable)) {
+            String outFile = "/tmp/index_rebuild_"+index+"_" + UUID.randomUUID().toString();
+            String indexName = SchemaUtil.getTableNameFromFullName(index);
+            conf.set(QueryServices.TRANSACTIONS_ENABLED, Boolean.TRUE.toString());
+            indexingTool.setConf(conf);
+            String [] args = {"-s", schema, "-dt", table, "-it", indexName, "-direct", "-runfg", "-op", outFile};
+            try {
+                LOGGER.info("Rebuilding index "+ index);
+                if(!dryRun) {
+                    indexingTool.run(args);
+                }
+            } catch (Exception e) {
+                LOGGER.severe("Something went wrong while building the index "+ index +" "+e);
+            }
+        }
+    }
+
+    private void extractTablesAndIndexes(PhoenixConnection conn) {
+        String [] tables = inputTables.trim().split(",");
+        PTable dataTable = null;
+        try {
+            for (String tableName : tables) {
+                HashSet<String> physicalIndexes = new HashSet<>();
+                dataTable = PhoenixRuntime.getTableNoCache(conn, tableName);
+                String physicalTableName = dataTable.getPhysicalName().getString();
+
+                for (PTable indexTable : dataTable.getIndexes()) {
```

Comment:
This feature is not supported on transactional tables. If a table is transactional, it should be skipped and this should be logged.

Reply:
Done